File: SudokuSolver_Kotlin/src/main/sudokusolver/kotlin/Rectangles.kt
package sudokusolver.kotlin

import java.util.EnumSet

class Rectangle(val cells: List<UnsolvedCell>) {
    val commonCandidates: EnumSet<SudokuNumber> = enumIntersect(*cells.map { it.candidates }.toTypedArray())

    val floor: List<UnsolvedCell> by lazy {
        cells.filter { it.candidates.size == 2 }
    }

    val roof: List<UnsolvedCell> by lazy {
        cells.filter { it.candidates.size > 2 }
    }
}

fun createRectangles(board: Board<Cell>): List<Rectangle> =
    board.rows
        .zipEveryPair()
        .flatMap { (rowA, rowB) ->
            (rowA zip rowB)
                .mapNotNull { (cellA, cellB) ->
                    if (cellA is UnsolvedCell && cellB is UnsolvedCell) cellA to cellB else null
                }
                .zipEveryPair()
                .map { (columnA, columnB) ->
                    Rectangle(listOf(columnA.first, columnA.second, columnB.first, columnB.second))
                }
        }
        .filter { rectangle ->
            rectangle.commonCandidates.size == 2 && rectangle.cells.map { it.block }.toSet().size == 2
        }

File: sudokusolver_scala/src/main/scala/sudokusolver/scala/Rectangles.scala
package sudokusolver.scala

class Rectangle(val cells: Seq[UnsolvedCell]):
  val commonCandidates: Set[SudokuNumber] = cells.map(_.candidates).reduce(_ & _)
  lazy val floor: Seq[UnsolvedCell] = cells.filter(_.candidates.size == 2)
  lazy val roof: Seq[UnsolvedCell] = cells.filter(_.candidates.size > 2)

def createRectangles(board: Board[Cell]): Seq[Rectangle] =
  board.rows
    .zipEveryPair
    .flatMap { (rowA, rowB) =>
      rowA.zip(rowB)
        .collect { case (cellA: UnsolvedCell, cellB: UnsolvedCell) => cellA -> cellB }
        .zipEveryPair
        .map((columnA, columnB) => Rectangle(Seq(columnA(0), columnA(1), columnB(0), columnB(1))))
    }
    .filter(rectangle => rectangle.commonCandidates.size == 2 && rectangle.cells.map(_.block).toSet.size == 2)